toml/src/main/kotlin/org/rust/toml/completion/crates/local/CratesLocalIndexService.kt

```diff
@@ -0,0 +1,212 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml.completion.crates.local
+
+import com.google.gson.Gson
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.components.service
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.Task
+import com.intellij.util.io.*
+import org.eclipse.jgit.lib.ObjectId
+import org.eclipse.jgit.lib.Repository
+import org.eclipse.jgit.revwalk.RevTree
+import org.eclipse.jgit.revwalk.RevWalk
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder
+import org.eclipse.jgit.treewalk.TreeWalk
+import org.rust.openapiext.pluginDirInSystem
+import org.rust.stdext.cleanDirectory
+import java.io.BufferedReader
+import java.io.DataInput
+import java.io.DataOutput
+import java.io.InputStreamReader
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.*
+
+class CratesLocalIndexService : Disposable {
+    private val baseCratesLocalRegistryDir: Path
+        get() = pluginDirInSystem().resolve("crates-local-index")
+
+    private val corruptionMarkerFile: Path
+        get() = baseCratesLocalRegistryDir.resolve(CORRUPTION_MARKER_NAME)
+
+    private val userCargoIndexDir: Path
+        get() = Paths.get(
+            System.getProperty("user.home"),
+            ".cargo/registry/index/github.com-1ecc6299db9ec823/.git/"
+        )
+
+    private val repository: Repository = FileRepositoryBuilder()
+        .setGitDir(userCargoIndexDir.toFile())
+        .build()
+
+    private var crates: PersistentHashMap<String, CargoRegistryCrate>
+
+    private val state: CratesLocalIndexState = service<CratesLocalIndexState>()
+
+    private var updateIsNeeded: Boolean = true
+
+    @Volatile
+    var isReady: Boolean = true
+
+    init {
+        checkInvalidatedStorage()
+
+        crates = PersistentHashMap(
+            baseCratesLocalRegistryDir.resolve("crates-local-index"),
+            EnumeratorStringDescriptor.INSTANCE,
+            CrateExternalizer
+        )
+
+        checkIfNeedUpdate()
+    }
+
+    private fun checkIfNeedUpdate() {
+        val branch = repository.resolve("origin/master") ?: return
+        val headCommitHash = branch.name
+
+        if (headCommitHash == state.indexedCommitHash) {
+            updateIsNeeded = false
+        }
+    }
+
+    private fun updateIfNeeded() {
+        if (updateIsNeeded) {
+            updateIsNeeded = false
+            isReady = false
+            CratesLocalIndexUpdateTask().queue()
+        }
+    }
+
+    fun checkInvalidatedStorage() {
```

Comment:
Probably something like this would be more clear because it's not just a check but also an invalidation action
```suggestion
    private fun invalidateStorageIfNeeded() {
```